cartex/src/runtime/mpi_comm/runtime.cpp

```diff
@@ -114,34 +138,331 @@ runtime::impl::neighborhood::neighborhood(
+            for (int x = 0; x < ext_buffer[0]; ++x) field[x + offset] = buffer_left[x + offset_l];
+        }
+        for (int y = 0; y < m_halos[3]; ++y)
+        {
+            const auto offset = ext_buffer[0] * (y + m_halos[2] + d.domain_ext[1] +
+                                                    (z + m_halos[4]) * ext_buffer[1]);
+            const auto offset_r = ext_buffer[0] * (y + m_halos[3] * z);
+            for (int x = 0; x < ext_buffer[0]; ++x) field[x + offset] = buffer_right[x + offset_r];
+        }
+    }
+#endif
+}
+
+void
+runtime::impl::neighborhood::exchange(memory_type& field, std::vector<memory_type>& send_buffers,
+    std::vector<memory_type>& recv_buffers, int field_id)
+{
+    pack_x(field, send_buffers[0], send_buffers[1]);
+    CARTEX_CHECK_MPI_RESULT(
+        MPI_Sendrecv(send_buffers[1].hd_data(), send_buffers[1].size() * sizeof(runtime::real_type),
+            MPI_BYTE, x_l.rank, sendtag(field_id, 0, true), recv_buffers[1].hd_data(),
+            recv_buffers[1].size() * sizeof(runtime::real_type), MPI_BYTE, x_r.rank,
+            recvtag(field_id, 0, true), comm, MPI_STATUS_IGNORE));
+    CARTEX_CHECK_MPI_RESULT(
+        MPI_Sendrecv(send_buffers[0].hd_data(), send_buffers[0].size() * sizeof(runtime::real_type),
+            MPI_BYTE, x_r.rank, sendtag(field_id, 0, false), recv_buffers[0].hd_data(),
+            recv_buffers[0].size() * sizeof(runtime::real_type), MPI_BYTE, x_l.rank,
+            recvtag(field_id, 0, false), comm, MPI_STATUS_IGNORE));
+    unpack_x(field, recv_buffers[0], recv_buffers[1]);
+
+    pack_y(field, send_buffers[2], send_buffers[3]);
+    CARTEX_CHECK_MPI_RESULT(
+        MPI_Sendrecv(send_buffers[3].hd_data(), send_buffers[3].size() * sizeof(runtime::real_type),
+            MPI_BYTE, y_l.rank, sendtag(field_id, 1, true), recv_buffers[3].hd_data(),
+            recv_buffers[3].size() * sizeof(runtime::real_type), MPI_BYTE, y_r.rank,
+            recvtag(field_id, 1, true), comm, MPI_STATUS_IGNORE));
+    CARTEX_CHECK_MPI_RESULT(
+        MPI_Sendrecv(send_buffers[2].hd_data(), send_buffers[2].size() * sizeof(runtime::real_type),
+            MPI_BYTE, y_r.rank, sendtag(field_id, 1, false), recv_buffers[2].hd_data(),
+            recv_buffers[2].size() * sizeof(runtime::real_type), MPI_BYTE, y_l.rank,
+            recvtag(field_id, 1, false), comm, MPI_STATUS_IGNORE));
+    unpack_y(field, recv_buffers[2], recv_buffers[3]);
+
+    const auto z0 = field.hd_data();
+    const auto z1 = field.hd_data() + z_plane * m_halos[4];
+    const auto z2 = field.hd_data() + z_plane * d.domain_ext[2];
+    const auto z3 = field.hd_data() + z_plane * (m_halos[4] + d.domain_ext[2]);
+    const auto left_size = z_plane * m_halos[5] * sizeof(runtime::real_type);
+    const auto right_size = z_plane * m_halos[4] * sizeof(runtime::real_type);
+    CARTEX_CHECK_MPI_RESULT(
+        MPI_Sendrecv(z1, left_size, MPI_BYTE, z_l.rank, sendtag(field_id, 2, true), z3, left_size,
```

Comment:
Just a comment - this will work only if there is no padding. With padding the Z-planes are no longer contiguous and still require (un)packing. I guess we don't have any padding in the benchmarks, but maybe we need some as yet another parameter to study it's impact on performance? ;P

Reply:
yes, correct, works only for non-padded fields. I will add a note, so we don't forget.